src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs

```diff
@@ -3,16 +3,214 @@
 
 using System.Collections.Generic;
 using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.CompilerServices;
 
 namespace System.Threading.Tasks
 {
     public static partial class Parallel
     {
-        /// <summary>Executes a for each operation on an <see cref="System.Collections.Generic.IEnumerable{TSource}"/> in which iterations may run in parallel.</summary>
+        /// <summary>Executes a for loop in which iterations may run in parallel.</summary>
+        /// <param name="fromInclusive">The start index, inclusive.</param>
+        /// <param name="toExclusive">The end index, exclusive.</param>
+        /// <param name="body">An asynchronous delegate that is invoked once per element in the data source.</param>
+        /// <exception cref="ArgumentNullException">The <paramref name="body"/> argument is <see langword="null"/>.</exception>
+        /// <returns>A task that represents the entire for each operation.</returns>
+        /// <remarks>The operation will execute at most <see cref="Environment.ProcessorCount"/> operations in parallel.</remarks>
+        public static Task ForAsync<T>(T fromInclusive, T toExclusive, Func<T, CancellationToken, ValueTask> body)
+            where T : notnull, IBinaryInteger<T>
+        {
+            if (fromInclusive is null) throw new ArgumentNullException(nameof(fromInclusive));
+            if (toExclusive is null) throw new ArgumentNullException(nameof(toExclusive));
+            ArgumentNullException.ThrowIfNull(body);
+
+            return ForAsync(fromInclusive, toExclusive, DefaultDegreeOfParallelism, TaskScheduler.Default, default, body);
+        }
+
+        /// <summary>Executes a for loop in which iterations may run in parallel.</summary>
+        /// <param name="fromInclusive">The start index, inclusive.</param>
+        /// <param name="toExclusive">The end index, exclusive.</param>
+        /// <param name="cancellationToken">A cancellation token that may be used to cancel the for each operation.</param>
+        /// <param name="body">An asynchronous delegate that is invoked once per element in the data source.</param>
+        /// <exception cref="ArgumentNullException">The <paramref name="body"/> argument is <see langword="null"/>.</exception>
+        /// <returns>A task that represents the entire for each operation.</returns>
+        /// <remarks>The operation will execute at most <see cref="Environment.ProcessorCount"/> operations in parallel.</remarks>
+        public static Task ForAsync<T>(T fromInclusive, T toExclusive, CancellationToken cancellationToken, Func<T, CancellationToken, ValueTask> body)
+            where T : notnull, IBinaryInteger<T>
+        {
+            if (fromInclusive is null) throw new ArgumentNullException(nameof(fromInclusive));
+            if (toExclusive is null) throw new ArgumentNullException(nameof(toExclusive));
```

Comment:
ThrowIfNull can't work over `T`?

Reply:
ThrowIfNull takes object.  You can pass a T, but then if it's a value type at best it's a nop and at worst you're relying on the JIT eliding the boxing, which it will typically do but won't be able to all the time, e.g. in tier 0 since those optimizations are disabled.  Since this T will 99.999% of the time be a value type and I'm not concerned about this method being inlinable, I just wrote it out manually.

Reply:
Should we consider creating an analyzer to suggest using the old null check + throw, instead of ThrowIfNull, for generic types?

I suspect the answer might be no, since you say 99.999% of the cases would be a value type.

Reply:
No, I think we want to encourage using ThrowIfNull in general.  This is just a special case.

We could have an analyzer for ThrowIfNull with known value types (either an actual value type or a generic constrained to be a value type).  There's already an issue for that.